drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+		    &i2c_esp32_driver_api);
+#endif /* CONFIG_I2C_0 */
+
+#ifdef CONFIG_I2C_1
+static const struct i2c_esp32_config i2c_esp32_config_1 = {
+	.sig = {
+		.sda_out = I2CEXT1_SDA_OUT_IDX,
+		.sda_in = I2CEXT1_SDA_IN_IDX,
+		.scl_out = I2CEXT1_SCL_OUT_IDX,
+		.scl_in = I2CEXT1_SCL_IN_IDX,
+	},
+	.pins = {
+		.scl = CONFIG_I2C_ESP32_1_SCL_PIN,
+		.sda = CONFIG_I2C_ESP32_1_SDA_PIN,
+		.sda_pull_up =
+			IS_ENABLED(CONFIG_I2C_ESP32_1_SDA_PULL_UP),
+		.scl_pull_up =
+			IS_ENABLED(CONFIG_I2C_ESP32_1_SCL_PULL_UP),
+		.default_level =
+			IS_ENABLED(CONFIG_I2C_ESP32_1_DEFAULT_LEVEL),
+	},
+	.enable_bits = {
+		.clk = DPORT_I2C_EXT1_CLK_EN,
+		.rst = DPORT_I2C_EXT1_RST,
+	},
+	.regs = {
+		.ctr = (u32_t *)I2C_CTR_REG(1),
+		.fifo_cfg = (u32_t *)I2C_FIFO_CONF_REG(1),
+		.fifo = (u32_t *)I2C_DATA_APB_REG(1),
+		.slave_addr = (u32_t *)I2C_SLAVE_ADDR_REG(1),
+		.intr = (u32_t *)I2C_INT_ENA_REG(1),
+		.intr_clr = (u32_t *)I2C_INT_CLR_REG(1),
+		.intr_status = (u32_t *)I2C_INT_STATUS_REG(1),
+		.intr_raw = (u32_t *)I2C_INT_RAW_REG(1),
+		.status = (u32_t *)I2C_SR_REG(1),
+		.cmd_base = (u32_t *)I2C_COMD0_REG(1),
+		.timeout = (u32_t *)I2C_TO_REG(1),
+		.scl_low_period = (u32_t *)I2C_SCL_LOW_PERIOD_REG(1),
+		.scl_high_period = (u32_t *)I2C_SCL_HIGH_PERIOD_REG(1),
+		.scl_start_h_time = (u32_t *)I2C_SCL_START_HOLD_REG(1),
+		.scl_rstart_s_time = (u32_t *)I2C_SCL_RSTART_SETUP_REG(1),
+		.scl_stop_h_time = (u32_t *)I2C_SCL_STOP_HOLD_REG(1),
+		.scl_stop_s_time = (u32_t *)I2C_SCL_STOP_SETUP_REG(1),
+		.sda_hold_time = (u32_t *)I2C_SDA_HOLD_REG(1),
+		.sda_sample_time = (u32_t *)I2C_SDA_SAMPLE_REG(1),
+	},
+	.mode = {
+		.tx_lsb_first =
+			IS_ENABLED(CONFIG_ESP32_I2C_1_TX_LSB_FIRST),
+		.rx_lsb_first =
+			IS_ENABLED(CONFIG_ESP32_I2C_1_RX_LSB_FIRST),
+	},
+	.irq = {
+		.source = ETS_I2C_EXT1_INTR_SOURCE,
+		.line = CONFIG_I2C_ESP32_1_IRQ,
+	},
+	.default_config = {
+		.raw = CONFIG_I2C_1_DEFAULT_CFG,
+	},
+};
+
+static struct i2c_esp32_data i2c_esp32_data_1;
+
+DEVICE_AND_API_INIT(i2c_esp32_1, CONFIG_I2C_1_NAME, &i2c_esp32_init,
+		    &i2c_esp32_data_1, &i2c_esp32_config_1,
+		    POST_KERNEL, CONFIG_I2C_INIT_PRIORITY,
+		    &i2c_esp32_driver_api);
+#endif /* CONFIG_I2C_1 */
+
+static int i2c_esp32_init(struct device *dev)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	struct i2c_esp32_data *data = dev->driver_data;
+	int key = irq_lock();
+
+	k_sem_init(&data->fifo_sem, 1, UINT_MAX);
+	k_sem_init(&data->transfer_sem, 1, UINT_MAX);
+
+	/* Initialize with the peripheral disabled */
+	i2c_esp32_set_device_enabled(config, false);
+
+	/* Initialize with the IRQ connected but disabled. */
+	irq_disable(config->irq.line);
+
+	/* Even if irq_enable() is called on config->irq.line, disable
+	 * interrupt sources in the I2C controller.
+	 */
+	*config->regs.intr = 0;
+	intr_matrix_set(0, config->irq.source, config->irq.line);
+
+	/* These if statements below are necessary as IRQ_CONNECT() requires
+	 * a constant argument.
```

Comment:
Not understanding the comment.  The expression "CONFIG_I2C_ESP32_0_IRQ" is going to be a constant for sure.  The variable in the config struct won't, but that's not what you're using.

Also: as mentioned before I really don't think these IRQs are actually configurable in any meaningful way.  I'm pretty sure they're constant at least within an SoC, which can be tested at compile time and removed from kconfig.

Reply:
This function is called twice: one for each I2C controller. Different lines are assigned for them.

The IRQ lines are configurable: what's fixed are how they're triggered (level, edge, etc), their priorities, and category (peripheral or internal).  If it's a peripheral interrupt line, sources can be routed (intr_matrix_set() call above).

Reply:
Ah, got it.  You want each "called" only once.

The dirty secret here though is that IRQ_CONNECT() doesn't actually produce any runtime code.  It's perfectly valid for you to set up both interrupts in this one function and "call" them twice at runtime.  Or do it from a different global init spot, etc...  All that needs to happen is that the compiler expands the macro only once.

Though thinking about it, the compiler will just eliminate the tests anyway as all its sees are empty blocks following the two tests.

Honestly IRQ_CONNECT() needs some rethinking as a framework.

Reply:
This can be implemented cleanly by defining an extra per-instance IRQ configuration function as described in Zephyr Device Drivers Documentation section [Single Driver, Multiple Instances](https://www.zephyrproject.org/doc/devices/drivers/drivers.html#single-driver-multiple-instances). The documentation calls this function 'config_func'. I find the name pretty confusing however, it is too general and there are already many config-s in the driver. In the device drivers I wrote I use always the name 'irq_config'.